[email] Added configurable email template

pandafy opened the pull request:
Fixes #2 
Blocked by #19 and #23 
While testing set
```
OPENWISP_NOTIFICATION_EMAIL_LOGO = 'https://git.io/Jfw82'
```
I have configured default setting to take image from master branch which is absent now.

pandafy commented:
Tested with monitoring module.

![Threshold crossed email](https://user-images.githubusercontent.com/32094433/82884949-19ff6200-9f62-11ea-81d9-0138e8d2855c.jpg)
![Device under threshold email](https://user-images.githubusercontent.com/32094433/82884116-ee2fac80-9f60-11ea-8cee-7595f961b88e.png)

NoumbissiValere commented on openwisp_notifications/tests/test_notifications.py:
```diff
@@ -346,3 +347,23 @@ def test_register_unregister_notification_type(self):
+    def test_notification_type_email(self):
+        self.notification_options.update({'type': 'default'})
+        self._create_notification()
+        # n = notification_queryset.first()
```
Please if this line no longer needed, could you remove it

pandafy commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,88 @@
+            <div class="notification-description">
+              <code> {{ notification.message }} </code>
+            </div>
+            <p>You can also view this notifification on the dashboard by following 
```
I suppose we don't need to provide translation files. If please tell me about which languages. :smile:

NoumbissiValere replied:
well, this is a simple scenario. an admin from `Peru` might get this module and his notifications will be supplied in English. we could reply on the translation of his email server but this not good as we are not sure what ever sever he will be using. in fact anyone from any part of the world who doesn't understand English could run this module.  loading `i18n` and `l10n` and using `trans` solves this problem for you.
Please indicate if you need more explanation

pandafy replied:
I was asking about `django.po` file. :sweat_smile: Do we need to provide that or assume that user will create and compile messages on their own?

NoumbissiValere replied:
Aahh. i didn't understand what you meant :smile: 
No you don't need to provide that file.

pandafy replied:
:+1:

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+              <p> {% trans 'Greetings' %} {{notification.recipient.username}},</p>
+              <p> {% trans 'You have a new notification from your network :' %} {{site}}</p>
+              <div class="notification-description">
+                <code> {{ notification.message }} </code>
```
Don't use code for this please, it's not code.

Also do not use preformatted text.

Use normal text with normal alignment (left).

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+      <table style="width: 100%;">
+        <thead class="notification-header">
+          <tr>
+            <th><img src="{{logo}}" alt="Organization Logo" id="org-logo" class="logo"></th>
```
it'd be nice if it was the organization logo (eg: each organization supplying their own logo), but it's not, so I would not use the words `org` and `Organization`

nemesisdesign commented on openwisp_notifications/settings.py:
```diff
@@ -3,6 +3,16 @@
+)
+
+OPENWISP_NOTIFICATION_EMAIL_LOGO = getattr(
+    settings, 'OPENWISP_NOTIFICATION_EMAIL_LOGO', 'https://git.io/JfwB3'
```
please make a note (eg: an issue) to update this once we merge it, so that we can change the default URL to point to the github URL.

pandafy replied:
This already points to the location on master branch:
https://raw.githubusercontent.com/openwisp/openwisp-notifications/master/openwisp_notifications/static/openwisp_notifications/images/openwisp-logo.png
I used git,io to shorten URL.

nemesisdesign replied:
@pandafy let's not shorten it please, users may have the impression we're hiding something. Please also add a point to the README which advises to change this setting to point to their own web server and not to github.

nemesisdesign commented on README.rst:
```diff
@@ -325,6 +326,61 @@ An example usage is shown below.
++-----------+--------------------------------------------------+
+|   type    |  ``str``                                         |
++-----------+--------------------------------------------------+
+|  default  |  `https://git.io/JfVhe` (OpenWISP logo)          |
```
same here

nemesisdesign commented on README.rst:
```diff
@@ -325,6 +326,61 @@ An example usage is shown below.
+See `openwisp_notifications/email_template.html <https://git.io/JfVpA>`_ for reference implementation.
+
+OPENWISP_NOTIFICATION_EMAIL_LOGO
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
use

```
``OPENWISP_NOTIFICATION_EMAIL_LOGO``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

nemesisdesign commented on README.rst:
```diff
@@ -325,6 +326,61 @@ An example usage is shown below.
+--------
+
+OPENWISP_NOTIFICATION_EMAIL_TEMPLATE
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
same here

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+  </div>
+  {% endblock body %}
+</body>
+
```
do not leave this blank line please

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+  <div class="background">
+    <div class="container">
+      <table style="width: 100%;">
+        <thead class="notification-header">
```
I would add a template block for this section

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+          </tr>
+        </thead>
+        <tbody>
+          <tr class="notification-body">
```
and this section too

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,93 @@
+          <tr class="notification-body">
+            <td colspan="2">
+              <p> {% trans 'Greetings' %} {{notification.recipient.username}},</p>
+              <p> {% trans 'You have a new notification from your network :' %} {{site}}</p>
```
do not leave spaces before punctuation characters...

pandafy commented:
Since we are using [send_mail](https://docs.djangoproject.com/en/3.0/topics/email/#send-mail) function and providing `message` argument as well. We need not worry about providing a setting for disabling HTML email since it will be fallback to value of `message` parameter which has content type of `text/plain`. Please let me know if my assumption in wrong. :grimacing: 

I have updated the code to strip down HTML tags from `message` property  of notification before passing it's content to `message` parameter of `send_mail`.

pandafy commented:
**Don't Assume It—Prove It**

With `message` property defined in notification type.
![Screenshot from 2020-06-03 01-27-16](https://user-images.githubusercontent.com/32094433/83563961-92a18800-a539-11ea-8ac3-11fa243d2102.png)

With `message template` property defined in notification type
![Screenshot from 2020-06-03 01-29-56](https://user-images.githubusercontent.com/32094433/83564181-e7dd9980-a539-11ea-8cb0-ba3829369207.png)


Tried with https://m.gmail.com which does not renders HTML emails.

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,95 @@
+          {% block notification_body %}
+          <tr class="notification-body">
+            <td colspan="2">
+              <p> {% trans 'Greetings' %} {{ notification.recipient.username }},</p>
```
don't leave spaces `<p> ` > `<p>`
apply to all the other lines please

nemesisdesign commented on openwisp_notifications/templates/openwisp_notifications/email_template.html:
```diff
@@ -0,0 +1,95 @@
+                <p> {{ notification.message }} </p>
+              </div>
+              <p>{% trans 'You can also view this notifification on the dashboard by following' %}
+                <a href="{{ site.domain }}{{ admin_url }}" id="admin-url">{% trans ' this link' %}.</a>
```
don't use "this link" and don't leave extra space.

I suggest building the URL outside of the template (like it's done for the plain text).

pandafy commented:
@nemesisdesign I mistakenly overwrote your merge from master branch due to force pushing. I have rebased on current master, I hope that fixes it. :frowning_face: 

Sorry for the trouble. :sweat:

pandafy commented:
Tested it with https://github.com/openwisp/openwisp-monitoring/pull/93/files
![Screenshot from 2020-06-06 03-56-51](https://user-images.githubusercontent.com/32094433/83927451-9851d480-a7aa-11ea-8cfd-d5da88778ba1.png)
![Screenshot from 2020-06-06 03-55-48](https://user-images.githubusercontent.com/32094433/83927457-99830180-a7aa-11ea-9c62-75198675db73.png)

pandafy commented:
Tested with master branch of openwisp-monitoring 
![Screenshot from 2020-06-06 04-18-41](https://user-images.githubusercontent.com/32094433/83928313-fa133e00-a7ac-11ea-9351-bc5c9e3283c9.png)
![Screenshot from 2020-06-06 04-13-51](https://user-images.githubusercontent.com/32094433/83928314-fb446b00-a7ac-11ea-9e1d-2a22cc36bbba.png)

pandafy commented:
Tested again with openwisp/openwisp-monitoring#93
Changes:
1. Email subject: Device name was coming twice, removed it (changed in openwisp/openwisp-monitoring#93)
2. Description: Removed `<br>` tag which was making notification description unnatural (changed in this PR)

![Screenshot from 2020-06-06 06-03-37](https://user-images.githubusercontent.com/32094433/83931793-326e4880-a7bc-11ea-9984-79f864f35f90.png)
![Screenshot from 2020-06-06 06-03-00](https://user-images.githubusercontent.com/32094433/83931796-34380c00-a7bc-11ea-991d-935757092a22.png)

pandafy commented:
Changes: Better UX for external link in HTML email

![Screenshot from 2020-06-06 17-33-24](https://user-images.githubusercontent.com/32094433/83943754-15696200-a81c-11ea-9934-50974c371425.png)
![Screenshot from 2020-06-06 17-32-33](https://user-images.githubusercontent.com/32094433/83943756-17332580-a81c-11ea-8ef4-1e22fb0ede8c.png)
